Add Support for Windows

### Expected Behavior

Add a script to generate a `.dll` file as part of build process. Also, edits needs to made in setup.py to include this generated dll file. This dll file will be generated using bazel build. 

### Current Behavior

Right now, pydp only supports Linux systems. This is due to the fact that we are just generating the `.so` file as seen [here](https://github.com/OpenMined/PyDP/blob/dev/setup.py#L40). 

### Logs

```
>>> import pydp as dp
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Anaconda3\lib\site-packages\pydp\__init__.py", line 1, in <module>
    from .pydp import *
ModuleNotFoundError: No module named 'pydp.pydp'
```

### System information

* PyDP version: 0.1.0dev
* Operating System: Any windows system

There are a few blockers with the windows build:

1. `copts = ["-Wno-sign-compare"],` are not supported on windows. This is a minor issue and we can simply remove them and it all works. I am unsure the best way to resolve this - if you use bazel on windows I'd love to hear ideas. 
2. The SHA that we are using includes [`#include <libgen.h>`](https://github.com/google/differential-privacy/blob/b7f4c39d9f73d67b34cdbd1b8483e5f72072fc73/cc/base/logging.cc#L22) in `base/logging.cc` and this is not supported on windows. We can remove it and upstream has already done so but we would need to pull in a newer "version" (SHA).
3. The last hurdle is that [upstream ](https://github.com/google/differential-privacy) are using [`#include <unistd>`](https://github.com/google/differential-privacy/blob/d9f9215deec7f6bc85b1076a220dc83795cb0be8/cc/base/logging.cc#L25) in `base/logging.cc` which is not supported on windows. If I remove it e get the following errors:
```
external/google_dp/base/logging.cc(79): error C3861: 'mkdir': identifier not found
external/google_dp/base/logging.cc(82): error C3861: 'S_ISDIR': identifier not found
external/google_dp/base/logging.cc(139): error C2065: 'F_OK': undeclared identifier
external/google_dp/base/logging.cc(139): error C3861: 'access': identifier not found
external/google_dp/base/logging.cc(140): error C2065: 'W_OK': undeclared identifier
external/google_dp/base/logging.cc(140): error C3861: 'access': identifier not found
external/google_dp/base/logging.cc(216): error C2065: 'CLOCK_REALTIME': undeclared identifier
external/google_dp/base/logging.cc(216): error C3861: 'clock_gettime': identifier not found
```
So we need to find or create windows alternatives for these and include them. As far as I am aware this would need an someone from the upstream project to complete. 

So 1) and 2) are not deal breakers but 3) seems to be a slight issue and will need assistance from [upstream](https://github.com/google/differential-privacy)

--- a/.bazelrc
+++ b/.bazelrc
@@ -1,1 +1,3 @@
-build --cxxopt='-std=c++17'
+build:Linux --cxxopt='-std=c++17'
+build:macOS --cxxopt='-std=c++17'
+build:Windows --cxxopt='/std:c++17'

--- a/.github/workflows/pypipublish_linux.yml
+++ b/.github/workflows/pypipublish_linux.yml
@@ -17,7 +17,7 @@ jobs:
         run: |
           git submodule update --init --recursive
           cd third_party/differential-privacy
-          git checkout 1b1dc6639173c0a13613189ec21851604a4c7335
+          git checkout 613c69ec6a7058c821b52ae6ad0ae04ad28270f1
           cd -
           rm -rf third_party/differential-privacy/java
           rm -rf third_party/differential-privacy/examples/java
@@ -54,7 +54,7 @@ jobs:
 
       - name: Build pydp lib
         run: |
-          bazel build src/python:bindings_test  --verbose_failures
+          bazel build --config Linux src/python:bindings_test  --verbose_failures
           rm -f _pydp.so
           cp -f ./bazel-bin/src/bindings/_pydp.so ./pydp
 

--- a/.github/workflows/pypipublish_osx.yml
+++ b/.github/workflows/pypipublish_osx.yml
@@ -17,7 +17,7 @@ jobs:
         run: |
           git submodule update --init --recursive
           cd third_party/differential-privacy
-          git checkout 1b1dc6639173c0a13613189ec21851604a4c7335
+          git checkout 613c69ec6a7058c821b52ae6ad0ae04ad28270f1
           cd -
           rm -rf third_party/differential-privacy/java
           rm -rf third_party/differential-privacy/examples/java
@@ -35,7 +35,7 @@ jobs:
 
       - name: Build pydp lib
         run: |
-          bazel build src/python:bindings_test  --verbose_failures
+          bazel build --config macOS src/python:bindings_test  --verbose_failures
           rm -f _pydp.so
           cp -f ./bazel-bin/src/bindings/_pydp.so ./pydp
 

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -10,7 +10,7 @@ jobs:
     runs-on: windows-latest
     strategy:
       matrix:
-        python-version: [3.6, 3.7, 3.8] # 3.5 Currently 3.5 does not run through the tests.yml CI. Leaving it out until testng is implemented.
+        python-version: [3.5, 3.6, 3.7, 3.8]
     steps:
     - uses: actions/checkout@v2
       with:

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -35,6 +35,8 @@ jobs:
          python setup.py bdist_wheel
          $version=(python -c "import pydp; print(pydp.__version__)")
          echo "::set-output name=version::$version"
+         $wheel_name=ls .\dist
+         echo "::set-output name=wheel_name::$wheel_name"
       id: build-pydp
     - name: Create Release
       id: create_release
@@ -43,7 +45,7 @@ jobs:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       with:
         tag_name: ${{ github.event.inputs.build_tag }}
-        release_name: "Release ${{ github.event.inputs.build_tag }}"
+        release_name: Release ${{ github.event.inputs.build_tag }}
         draft: false
         prerelease: true
     - name: Upload Release Asset
@@ -53,8 +55,8 @@ jobs:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       with:
         upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps 
-        asset_path: .\dist\*.whl
-        asset_name: "python_dp-${{ github.event.inputs.build_tag }}.whl"
+        asset_path: ${{steps.build-pydp.outputs.wheel_name}}
+        asset_name: python_dp-${{ github.event.inputs.build_tag }}.whl
         asset_content_type: application/zip
   
         # - name: Build and publish

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -40,7 +40,9 @@ jobs:
          echo "::set-output name=version::$version"
          $wheel_path=ls .\dist
          echo "::set-output name=wheel_path::$wheel_path"
-         echo "::set-output name=wheel_name::$wheel_path.name"
+         $wheel_name=(ls .\dist).name
+         echo "::set-output name=wheel_name::$wheel_name"
+
 
       id: build-pydp
     - name: Test the wheel

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -1,12 +1,8 @@
 name: Windows Package github deploy
 
-on: 
-  workflow_dispatch:
-    inputs:
-      build_tag:
-        description: 'The tag and version of the release.'     
-        required: true
-        default: '0.x.x-win_amd64-beta.x'
+on:
+  release:
+    types: [published]
 
 jobs:
   build-windows-wheel:
@@ -14,7 +10,7 @@ jobs:
     runs-on: windows-latest
     strategy:
       matrix:
-        python-version: [3.5, 3.6, 3.7, 3.8]
+        python-version: [3.6, 3.7, 3.8] # 3.5 Currently 3.5 does not run through the tests.yml CI. Leaving it out until testng is implemented.
     steps:
     - uses: actions/checkout@v2
       with:
@@ -49,32 +45,12 @@ jobs:
       run: |
           python -m pip install ${{steps.build-pydp.outputs.wheel_path}}
           python -m pip install pytest
-          pytest tests/  
-    - name: Create Release
-      id: create_release
-      uses: actions/create-release@v1
-      env:
-        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-      with:
-        tag_name: ${{ matrix.python-version }}-${{ github.event.inputs.build_tag }}
-        release_name: Release ${{ matrix.python-version }}-${{ github.event.inputs.build_tag }}
-        draft: false
-        prerelease: true
-    - name: Upload Release Asset
-      id: upload-release-asset 
-      uses: actions/upload-release-asset@v1
-      env:
-        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-      with:
-        upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps 
-        asset_path: ${{steps.build-pydp.outputs.wheel_path}}
-        asset_name: ${{ steps.build-pydp.outputs.wheel_name}}
-        asset_content_type: application/zip
-  
-        # - name: Build and publish
-    #   env:
-    #     TWINE_USERNAME: ${{ secrets.PYPI_USERNAME }}
-    #     TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
-    #   run: |
-    #     pip wheel . -w wheel/
-    #     twine upload --skip-existing wheel/*
+          pytest tests/
+          twine check ${{steps.build-pydp.outputs.wheel_path}}  
+    
+    - name: Publishing the wheel
+        env:
+          TWINE_USERNAME: __token__
+          TWINE_PASSWORD: ${{ secrets.TOKEN }}
+        run: |
+          twine upload --skip-existing ${{steps.build-pydp.outputs.wheel_path}}

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -1,4 +1,4 @@
-name: Windows Package
+name: Windows Package github deploy
 
 on: 
   workflow_dispatch:

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -35,6 +35,8 @@ jobs:
          python setup.py bdist_wheel
          $version=(python -c "import pydp; print(pydp.__version__)")
          echo "::set-output name=version::$version"
+         $wheel_name=ls .\dist
+         echo "::set-output name=wheel_name::$wheel_name.name"
       id: build-pydp
     - name: Create Release
       id: create_release
@@ -43,7 +45,7 @@ jobs:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       with:
         tag_name: ${{ github.event.inputs.build_tag }}
-        release_name: "Release ${{ github.event.inputs.build_tag }}"
+        release_name: Release ${{ github.event.inputs.build_tag }}
         draft: false
         prerelease: true
     - name: Upload Release Asset
@@ -53,8 +55,8 @@ jobs:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       with:
         upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps 
-        asset_path: .\dist\*.whl
-        asset_name: "python_dp-${{ github.event.inputs.build_tag }}.whl"
+        asset_path: .\dist\${{steps.build-pydp.outputs.wheel_name}}
+        asset_name: python_dp-${{ github.event.inputs.build_tag }}.whl
         asset_content_type: application/zip
   
         # - name: Build and publish

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -32,7 +32,7 @@ jobs:
       run: |
          bazel build --cxxopt='/std:c++17' src/python:bindings_test --verbose_failures
          copy bazel-bin\src\bindings\_pydp.so pydp\_pydp.pyd
-         python3.8 setup.py bdist_wheel
+         python setup.py bdist_wheel
          $version=(python -c "import pydp; print(pydp.__version__)")
          echo "::set-output name=version::$version"
          $wheel_name=ls .\dist

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -6,24 +6,27 @@ on:
       build_tag:
         description: 'The tag and version of the release.'     
         required: true
-        default: '0.x.x-cp38-cp38-win_amd64-beta.x'
+        default: '0.x.x-win_amd64-beta.x'
 
 jobs:
   build-windows-wheel:
 
     runs-on: windows-latest
+    strategy:
+      matrix:
+        python-version: [3.5, 3.6, 3.7, 3.8]
     steps:
     - uses: actions/checkout@v2
       with:
           submodules: true
-    - name: Set up Python
+    - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v1
       with:
-        python-version: 3.8
+        python-version: ${{ matrix.python-version }}
     - name: Install dependencies
       run: |
-        python3.8 -m pip install --upgrade pip
-        python3.8 -m pip install setuptools wheel twine auditwheel
+        python -m pip install --upgrade pip
+        python -m pip install setuptools wheel twine auditwheel
     - name: Setup msbuild
       uses: microsoft/setup-msbuild@v1.0.0
     - name: Setup Bazel
@@ -56,7 +59,7 @@ jobs:
       with:
         upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps 
         asset_path: ${{steps.build-pydp.outputs.wheel_name}}
-        asset_name: python_dp-${{ github.event.inputs.build_tag }}.whl
+        asset_name: ${{ steps.build-pydp.outputs.wheel_name.name }}
         asset_content_type: application/zip
   
         # - name: Build and publish

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -38,9 +38,16 @@ jobs:
          python setup.py bdist_wheel
          $version=(python -c "import pydp; print(pydp.__version__)")
          echo "::set-output name=version::$version"
-         $wheel_name=ls .\dist
-         echo "::set-output name=wheel_name::$wheel_name"
+         $wheel_path=ls .\dist
+         echo "::set-output name=wheel_path::$wheel_path"
+         echo "::set-output name=wheel_name::$wheel_path.name"
+
       id: build-pydp
+    - name: Test the wheel
+      run: |
+          python -m pip install ${{steps.build-pydp.outputs.wheel_path}}
+          python -m pip install pytest
+          pytest tests/  
     - name: Create Release
       id: create_release
       uses: actions/create-release@v1
@@ -58,8 +65,8 @@ jobs:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       with:
         upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps 
-        asset_path: ${{steps.build-pydp.outputs.wheel_name}}
-        asset_name: ${{ steps.build-pydp.outputs.wheel_name.name }}
+        asset_path: ${{steps.build-pydp.outputs.wheel_path}}
+        asset_name: ${{ steps.build-pydp.outputs.wheel_name}}
         asset_content_type: application/zip
   
         # - name: Build and publish

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -1,0 +1,66 @@
+name: Windows Package
+
+on: 
+  workflow_dispatch:
+    inputs:
+      build_tag:
+        description: 'The tag and version of the release.'     
+        required: true
+        default: '0.x.x-cp38-win_amd64-beta.x'
+
+jobs:
+  build-windows-wheel:
+
+    runs-on: windows-latest
+    steps:
+    - uses: actions/checkout@v2
+      with:
+          submodules: true
+    - name: Set up Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.8
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install setuptools wheel twine auditwheel
+    - name: Setup msbuild
+      uses: microsoft/setup-msbuild@v1.0.0
+    - name: Setup Bazel
+      uses: abhinavsingh/setup-bazel@v3
+    - name: Build PyDP
+      run: |
+         bazel build  src/python:bindings --verbose_failures
+         copy bazel-bin\src\bindings\_pydp.so pydp\_pydp.pyd
+         python setup.py bdist_wheel
+         $version=(python -c "import pydp; print(pydp.__version__)")
+         echo "::set-output name=version::$version"
+      id: build-pydp
+    - name: Create Release
+      id: create_release
+      uses: actions/create-release@v1
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      with:
+        tag_name: ${{ github.event.inputs.build_tag }}
+        release_name: "Release ${{ github.event.inputs.build_tag }}"
+        draft: false
+        prerelease: true
+    - name: Upload Release Asset
+      id: upload-release-asset 
+      uses: actions/upload-release-asset@v1
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      with:
+        upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps 
+        asset_path: .\dist\*.whl
+        asset_name: "python_dp-${{ github.event.inputs.build_tag }}.whl"
+        asset_content_type: application/zip
+  
+        # - name: Build and publish
+    #   env:
+    #     TWINE_USERNAME: ${{ secrets.PYPI_USERNAME }}
+    #     TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
+    #   run: |
+    #     pip wheel . -w wheel/
+    #     twine upload --skip-existing wheel/*

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -30,7 +30,7 @@ jobs:
       uses: abhinavsingh/setup-bazel@v3
     - name: Build PyDP
       run: |
-         bazel build  src/python:bindings_test --verbose_failures
+         bazel build --cxxopt='/std:c++17' src/python:bindings_test --verbose_failures
          copy bazel-bin\src\bindings\_pydp.so pydp\_pydp.pyd
          python setup.py bdist_wheel
          $version=(python -c "import pydp; print(pydp.__version__)")

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -49,8 +49,8 @@ jobs:
           twine check ${{steps.build-pydp.outputs.wheel_path}}  
     
     - name: Publishing the wheel
-        env:
-          TWINE_USERNAME: __token__
-          TWINE_PASSWORD: ${{ secrets.TOKEN }}
-        run: |
-          twine upload --skip-existing ${{steps.build-pydp.outputs.wheel_path}}
+      env:
+        TWINE_USERNAME: __token__
+        TWINE_PASSWORD: ${{ secrets.TOKEN }}
+      run: |
+        twine upload --skip-existing ${{steps.build-pydp.outputs.wheel_path}}

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -30,7 +30,7 @@ jobs:
       uses: abhinavsingh/setup-bazel@v3
     - name: Build PyDP
       run: |
-         bazel build  src/python:bindings --verbose_failures
+         bazel build  src/python:bindings_test --verbose_failures
          copy bazel-bin\src\bindings\_pydp.so pydp\_pydp.pyd
          python setup.py bdist_wheel
          $version=(python -c "import pydp; print(pydp.__version__)")

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -12,14 +12,17 @@ jobs:
   build-windows-wheel:
 
     runs-on: windows-latest
+    strategy:
+      matrix:
+        python-version: [3.5, 3.6, 3.7, 3.8]
     steps:
     - uses: actions/checkout@v2
       with:
           submodules: true
-    - name: Set up Python
+    - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v1
       with:
-        python-version: 3.8
+        python-version: ${{ matrix.python-version }}
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
@@ -56,7 +59,7 @@ jobs:
       with:
         upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps 
         asset_path: ${{steps.build-pydp.outputs.wheel_name}}
-        asset_name: python_dp-${{ github.event.inputs.build_tag }}.whl
+        asset_name: ${{ steps.build-pydp.outputs.wheel_name.name }}
         asset_content_type: application/zip
   
         # - name: Build and publish

--- a/.github/workflows/pypipublish_windows.yml
+++ b/.github/workflows/pypipublish_windows.yml
@@ -54,8 +54,8 @@ jobs:
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       with:
-        tag_name: ${{ github.event.inputs.build_tag }}
-        release_name: Release ${{ github.event.inputs.build_tag }}
+        tag_name: ${{ matrix.python-version }}-${{ github.event.inputs.build_tag }}
+        release_name: Release ${{ matrix.python-version }}-${{ github.event.inputs.build_tag }}
         draft: false
         prerelease: true
     - name: Upload Release Asset

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -37,7 +37,10 @@ jobs:
       uses: abhinavsingh/setup-bazel@v3
     - name: Build PyDP with coverage
     # removed c++ coverage for now 
+    # --expunge is needed as there is an intermittent error on macOS for which this seems to be the solution
+    # Ref: https://github.com/bazelbuild/rules_go/issues/2221  
       run: |
+        bazel clean --expunge 
         bazel build --config ${{runner.os}} src/python:bindings_test --verbose_failures    
     - name: Copy binary to python package(windows)  
       if: runner.os == 'Windows'

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -15,7 +15,7 @@ jobs:
       fail-fast: false
       # max-parallel: 6
       matrix:
-        os: [ubuntu-20.04, macOS-10.15, windows-2019]
+        os: [windows-2019] # [ubuntu-20.04, macOS-10.15, 
         python-version:  [3.6, 3.7, 3.8] # 3.5 fails as pipenv installs black as a dev dep for which 3.5 is not supported
 
     # Timeout: https://stackoverflow.com/a/59076067/4521646

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -3,7 +3,7 @@ name: Tests
 on:
   push:
     branches:
-    - master, dev
+    - windows-build
   pull_request:
     types: [opened, synchronize, reopened]
 
@@ -53,6 +53,8 @@ jobs:
         python -m pip install pipenv
         pipenv install setuptools wheel pytest gcovr coverage twine
         pipenv run python setup.py bdist_wheel
+    - name: Check the wheel
+      run: |
         pipenv run twine check dist/*
     - name: Install PyDP
       if: runner.os != 'Windows'

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -3,7 +3,7 @@ name: Tests
 on:
   push:
     branches:
-    - windows-build
+    - master, dev
   pull_request:
     types: [opened, synchronize, reopened]
 

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -15,7 +15,7 @@ jobs:
       fail-fast: false
       # max-parallel: 6
       matrix:
-        os: [windows-2019] # [ubuntu-20.04, macOS-10.15, 
+        os: [windows-2019, ubuntu-20.04, macOS-10.15] 
         python-version:  [3.6, 3.7, 3.8] # 3.5 fails as pipenv installs black as a dev dep for which 3.5 is not supported
 
     # Timeout: https://stackoverflow.com/a/59076067/4521646

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -29,10 +29,10 @@ jobs:
       uses: actions/setup-python@v1
       with:
         python-version: ${{ matrix.python-version }}
-    # - name: Set environment variables
-    #   uses: allenevans/set-env@v1.0.0
-    #   with:
-    #     PIPENV_DEFAULT_PYTHON_VERSION: ${{ matrix.python-version }}
+    - name: Set environment variables
+      uses: allenevans/set-env@v1.0.0
+      with:
+        PIPENV_DEFAULT_PYTHON_VERSION: ${{ matrix.python-version }}
     - name: Setup Bazel
       uses: abhinavsingh/setup-bazel@v3
     - name: Build PyDP with coverage

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -48,13 +48,19 @@ jobs:
       run: |
         cp -f ./bazel-bin/src/bindings/_pydp.so ./pydp  
           
-    - name: Setup and Install PyDP
+    - name: Setup and Build PyDP
       run: |
         python -m pip install pipenv
         pipenv install setuptools wheel pytest gcovr coverage twine
         pipenv run python setup.py bdist_wheel
         pipenv run twine check dist/*
+    - name: Install PyDP
         pipenv run python -m pip install --force-reinstall dist/*.whl
+    - name: Install PyDP (windows)
+      if: runner.os == 'Windows'
+      run: |
+         $wheel_path=ls .\dist
+         pipenv run python -m pip install --force-reinstall ${{wheel_path}}
     - name: Run tests
       run: |
         pipenv run coverage run -m pytest tests

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -61,8 +61,8 @@ jobs:
     - name: Install PyDP (windows)
       if: runner.os == 'Windows'
       run: |
-         wheel_path=ls .\dist
-         pipenv run python -m pip install --force-reinstall ${{wheel_path}}
+         $wheel_path=ls .\dist
+         pipenv run python -m pip install --force-reinstall $wheel_path
     - name: Run tests
       run: |
         pipenv run coverage run -m pytest tests

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -15,8 +15,8 @@ jobs:
       fail-fast: false
       # max-parallel: 6
       matrix:
-        os: [windows-2019] #, ubuntu-20.04, macOS-10.15] 
-        python-version:  [3.5] #, 3.6, 3.7, 3.8] 
+        os: [windows-2019, ubuntu-20.04, macOS-10.15] 
+        python-version:  [3.5, 3.6, 3.7, 3.8] 
 
     # Timeout: https://stackoverflow.com/a/59076067/4521646
     timeout-minutes: 35

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -61,7 +61,7 @@ jobs:
     - name: Install PyDP (windows)
       if: runner.os == 'Windows'
       run: |
-         $wheel_path=ls .\dist
+         wheel_path=ls .\dist
          pipenv run python -m pip install --force-reinstall ${{wheel_path}}
     - name: Run tests
       run: |

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -55,6 +55,8 @@ jobs:
         pipenv run python setup.py bdist_wheel
         pipenv run twine check dist/*
     - name: Install PyDP
+      if: runner.os != 'Windows'
+      run: |
         pipenv run python -m pip install --force-reinstall dist/*.whl
     - name: Install PyDP (windows)
       if: runner.os == 'Windows'

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -3,7 +3,7 @@ name: Tests
 on:
   push:
     branches:
-    - master, dev
+    - windows-build
   pull_request:
     types: [opened, synchronize, reopened]
 
@@ -15,8 +15,8 @@ jobs:
       fail-fast: false
       # max-parallel: 6
       matrix:
-        os: [windows-2019, ubuntu-20.04, macOS-10.15] 
-        python-version:  [3.6, 3.7, 3.8] # 3.5 fails as pipenv installs black as a dev dep for which 3.5 is not supported
+        os: [windows-2019] #, ubuntu-20.04, macOS-10.15] 
+        python-version:  [3.5] #, 3.6, 3.7, 3.8] 
 
     # Timeout: https://stackoverflow.com/a/59076067/4521646
     timeout-minutes: 35
@@ -54,7 +54,7 @@ jobs:
     - name: Setup and Build PyDP
       run: |
         python -m pip install pipenv
-        pipenv install setuptools wheel pytest gcovr coverage twine
+        pipenv install setuptools wheel pytest gcovr coverage twine --ignore-pipfile --skip-lock
         pipenv run python setup.py bdist_wheel
     - name: Check the wheel
       run: |

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -3,42 +3,69 @@ name: Tests
 on:
   push:
     branches:
-    - master, dev
+    - windows-build
   pull_request:
     types: [opened, synchronize, reopened]
 
 jobs:
-  build:
+  pytest:
 
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
     strategy:
-      max-parallel: 4
+      fail-fast: false
+      # max-parallel: 6
       matrix:
-        python-version: [3.5, 3.6, 3.7, 3.8]
+        os: [ubuntu-20.04, macOS-10.15, windows-2019]
+        python-version:  [3.6, 3.7, 3.8] # 3.5 fails as pipenv installs black as a dev dep for which 3.5 is not supported
+
+    # Timeout: https://stackoverflow.com/a/59076067/4521646
+    timeout-minutes: 35
 
     steps:
     - uses: actions/checkout@v2
+      with:
+          submodules: true
     - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v1
       with:
         python-version: ${{ matrix.python-version }}
-    - name: Set environment variables
-      uses: allenevans/set-env@v1.0.0
-      with:
-        PIPENV_DEFAULT_PYTHON_VERSION: ${{ matrix.python-version }}
-    - name: Install pre-requisites
+    # - name: Set environment variables
+    #   uses: allenevans/set-env@v1.0.0
+    #   with:
+    #     PIPENV_DEFAULT_PYTHON_VERSION: ${{ matrix.python-version }}
+    - name: Setup Bazel
+      uses: abhinavsingh/setup-bazel@v3
+    - name: Build PyDP with coverage
+    # removed c++ coverage for now 
       run: |
-        sudo apt-get update
-        bash ./prereqs_linux.sh
-    - name: Build pydp lib
+        bazel build --config ${{runner.os}} src/python:bindings_test --verbose_failures    
+    - name: Copy binary to python package(windows)  
+      if: runner.os == 'Windows'
       run: |
-        bash ./build_PyDP.sh
+        copy bazel-bin\src\bindings\_pydp.so pydp\_pydp.pyd
+    - name: Copy binary to python package   
+      if: runner.os != 'Windows'
+      run: |
+        cp -f ./bazel-bin/src/bindings/_pydp.so ./pydp  
+          
+    - name: Setup and Install PyDP
+      run: |
+        python -m pip install pipenv
+        pipenv install setuptools wheel pytest gcovr coverage twine
+        pipenv run python setup.py bdist_wheel
+        pipenv run twine check dist/*
+        pipenv run python -m pip install --force-reinstall dist/*.whl
     - name: Run tests
       run: |
-        make run-tests-only
+        pipenv run coverage run -m pytest tests
     - name: Check Python code coverage
+      if: runner.os == 'Linux' # Coverage will be the same on all systems so only running the check on Linux
       run: |
         make check-coverage-python
-    - name: Check C++ code coverage
-      run: |
-        make check-coverage-cpp
+      env:
+        MIN_COVERAGE: 80
+    # Need to see how to run the c++ code coverage
+    # - name: Check C++ code coverage
+    #   if: runner.os == 'Linux' # Coverage will be the same on all systems so only running the check on Linux
+    #   run: |
+    #     make check-coverage-cpp

--- a/.gitmodules
+++ b/.gitmodules
@@ -1,3 +1,3 @@
 [submodule "third_party/differential-privacy"]
 	path = third_party/differential-privacy
-	url = https://github.com/google/differential-privacy
+	url = https://github.com/benjaminDev/differential-privacy@windows-support-continued

--- a/.gitmodules
+++ b/.gitmodules
@@ -1,3 +1,3 @@
 [submodule "third_party/differential-privacy"]
 	path = third_party/differential-privacy
-	url = https://github.com/benjaminDev/differential-privacy
+	url = https://github.com/google/differential-privacy

--- a/.gitmodules
+++ b/.gitmodules
@@ -1,3 +1,3 @@
 [submodule "third_party/differential-privacy"]
 	path = third_party/differential-privacy
-	url = https://github.com/benjaminDev/differential-privacy@windows-support-continued
+	url = https://github.com/benjaminDev/differential-privacy

--- a/Dockerfile
+++ b/Dockerfile
@@ -12,7 +12,7 @@ ARG BAZEL_DOWNLOAD_URL=https://github.com/bazelbuild/bazel/releases/download
 ENV HOME=/root
 ENV PROJECT_DIR="${HOME}/PyDP"
 ENV PATH="/root/bin:${PATH}"
-ENV DP_SHA="1b1dc6639173c0a13613189ec21851604a4c7335"
+ENV DP_SHA="613c69ec6a7058c821b52ae6ad0ae04ad28270f1"
 
 # Define working directory
 WORKDIR ${HOME}
@@ -62,12 +62,12 @@ ENV PIPENV_VENV_IN_PROJECT=true
 
 # Build the bindings using Bazel and create a python wheel
 RUN pipenv --python ${PYTHON_VERSION} && \
-    pipenv run bazel build src/python:bindings_test  --verbose_failures
+    pipenv run bazel build --config Linux src/python:bindings_test  --verbose_failures
 
 RUN cp -f ./bazel-bin/src/bindings/_pydp.so ./pydp && \
     rm -rf dist/ && \
     pipenv run python setup.py bdist_wheel && \
-    pipenv install dist/*.whl 
+    pipenv install dist/*.whl
 
 # This `activates` the virtual env
 ENV VIRTUAL_ENV=$PROJECT_DIR/.venv

--- a/prereqs_linux.sh
+++ b/prereqs_linux.sh
@@ -59,7 +59,7 @@ git submodule update --init --recursive
 
 
 # checkout out to particular commit
-cd third_party/differential-privacy && git checkout 1b1dc6639173c0a13613189ec21851604a4c7335 && \
+cd third_party/differential-privacy && git checkout 613c69ec6a7058c821b52ae6ad0ae04ad28270f1 && \
 cd -
 # renaming workspace.bazel to workspace
 mv third_party/differential-privacy/cc/WORKSPACE.bazel third_party/differential-privacy/cc/WORKSPACE

--- a/prereqs_mac.sh
+++ b/prereqs_mac.sh
@@ -49,7 +49,7 @@ fi
 git submodule update --init --recursive
 
 # checkout out to particular commit
-cd third_party/differential-privacy && git checkout 1b1dc6639173c0a13613189ec21851604a4c7335
+cd third_party/differential-privacy && git checkout 613c69ec6a7058c821b52ae6ad0ae04ad28270f1
 
 cd -
 # renaming workspace.bazel to workspace

--- a/setup.py
+++ b/setup.py
@@ -44,6 +44,7 @@ setup(
     install_requires=requirements,
     license="Apache-2.0",
     long_description=read("docs/readme.rst"),
+    long_description_content_type='text/x-rst',
     include_package_data=True,
     keywords="pydp",
     name="python-dp",

--- a/setup.py
+++ b/setup.py
@@ -21,6 +21,7 @@ def read(fname):
     with open(os.path.join(os.path.dirname(__file__), fname), encoding="utf-8") as fp:
         return fp.read()
 
+
 requirements = []
 
 setup_requirements = []
@@ -44,11 +45,11 @@ setup(
     install_requires=requirements,
     license="Apache-2.0",
     long_description=read("docs/readme.rst"),
-    long_description_content_type='text/x-rst',
+    long_description_content_type="text/x-rst",
     include_package_data=True,
     keywords="pydp",
     name="python-dp",
-    package_data={"pydp": ["_pydp.so","_pydp.pyd"],},
+    package_data={"pydp": ["_pydp.so", "_pydp.pyd"],},
     packages=find_packages(exclude=["tests"]),  # need to check this
     setup_requires=setup_requirements,
     test_suite="tests",

--- a/setup.py
+++ b/setup.py
@@ -18,8 +18,8 @@ class BinaryDistribution(Distribution):
 
 
 def read(fname):
-    return open(os.path.join(os.path.dirname(__file__), fname)).read()
-
+    with open(os.path.join(os.path.dirname(__file__), fname), encoding="utf-8") as fp:
+        return fp.read()
 
 requirements = []
 

--- a/setup.py
+++ b/setup.py
@@ -47,7 +47,7 @@ setup(
     include_package_data=True,
     keywords="pydp",
     name="python-dp",
-    package_data={"pydp": ["_pydp.so"],},
+    package_data={"pydp": ["_pydp.so","_pydp.pyd"],},
     packages=find_packages(exclude=["tests"]),  # need to check this
     setup_requires=setup_requirements,
     test_suite="tests",

